src/core/lv2/Lv2Ports.cpp

```diff
@@ -0,0 +1,259 @@
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+
+#include "Lv2Ports.h"
+
+#ifdef LMMS_HAVE_LV2
+
+#include "Engine.h"
+#include "Lv2Basics.h"
+#include "Lv2Manager.h"
+
+namespace Lv2Ports {
+
+
+
+
+const char *toStr(Flow pf)
+{
+	switch(pf)
+	{
+		case Flow::Unknown: return "unknown";
+		case Flow::Input: return "input";
+		case Flow::Output: return "output";
+	}
+	return "";
+}
+
+
+
+
+const char *toStr(Type pt)
+{
+	switch(pt)
+	{
+		case Type::Unknown: return "unknown";
+		case Type::Control: return "control";
+		case Type::Audio: return "audio";
+		case Type::Event: return "event";
+		case Type::Cv: return "cv";
+	}
+	return "";
+}
+
+
+
+
+const char *toStr(Vis pv)
+{
+	switch(pv)
+	{
+		case Vis::Toggled: return "toggled";
+		case Vis::Enumeration: return "enumeration";
+		case Vis::Integer: return "integer";
+		case Vis::None: return "none";
+	}
+	return "";
+}
+
+
+
+
+std::vector<PluginIssue> Meta::get(const LilvPlugin *plugin,
+										unsigned int portNum)
+{
+	std::vector<PluginIssue> portIssues;
+	auto issue = [&portIssues](PluginIssueType i, const char* msg = "") {
+		portIssues.emplace_back(i, msg); };
+
+	Lv2Manager* man = Engine::getLv2Manager();
+	AutoLilvNode connectionOptional = man->uri(LV2_CORE__connectionOptional);
+
+	const LilvPort* lilvPort = lilv_plugin_get_port_by_index(plugin, portNum);
+
+	auto portFunc = [&plugin, &lilvPort, &man](
+		bool (*fptr)(const LilvPlugin*, const LilvPort*, const LilvNode*),
+		const char* str) {
+		bool res = fptr(plugin, lilvPort, man->uri(str).get());
+		return res;
+	};
+
+	auto hasProperty = [&portFunc](const char* str) {
+		return portFunc(lilv_port_has_property, str); };
+	auto isA = [&portFunc](const char* str) {
+		return portFunc(lilv_port_is_a, str); };
+
+
+	const char* portName = lilv_node_as_string(lilv_port_get_name(plugin, lilvPort));
+
+	m_optional = lilv_port_has_property(plugin, lilvPort,
+										connectionOptional.get());
+
+	m_vis = hasProperty(LV2_CORE__integer)
+		? Vis::Integer // WARNING: this may still be changed below
+		: hasProperty(LV2_CORE__enumeration)
+		? Vis::Enumeration
+		: hasProperty(LV2_CORE__toggled)
+		? Vis::Toggled
+		: Vis::None;
+
+	if (isA(LV2_CORE__InputPort)) {
+		m_flow = Flow::Input;
+	}
+	else if (isA(LV2_CORE__OutputPort)) {
+		m_flow = Flow::Output;
+	} else {
+		m_flow = Flow::Unknown;
+		issue(unknownPortFlow, portName);
+	}
+
+	m_def = .0f; m_min = .0f; m_max = .0f;
+
+	if (isA(LV2_CORE__ControlPort)) {
+		m_type = Type::Control;
+
+		if(m_flow == Flow::Input)
+		{
+			bool isToggle = m_vis == Vis::Toggled;
+
+			LilvNode * defN, * minN = nullptr, * maxN = nullptr;
+			lilv_port_get_range(plugin, lilvPort, &defN,
+					isToggle ? nullptr : &minN,
+					isToggle ? nullptr : &maxN);
+
+			auto takeRangeValue = [&](LilvNode* node,
+				float& storeHere, PluginIssueType it) {
+				if(node) {
```

Comment:
More explicit blocking, also in lines below this
```suggestion
				if(node)
				{
```